[CSS Zoom] Apply zoom to Preferred, Minimum, and Maximum sizes when evaluating

sammygill opened the pull request:
#### 49df8515af2f2669b932dafb26b26d7d2c5e5dfd
<pre>
[CSS Zoom] Apply zoom to Preferred, Minimum, and Maximum sizes when evaluating
<a href="https://bugs.webkit.org/show_bug.cgi?id=300052">https://bugs.webkit.org/show_bug.cgi?id=300052</a>
<a href="https://rdar.apple.com/161848512">rdar://161848512</a>

Reviewed by Tim Nguyen.

This patch marks the Preferred, Minimum, Maximum sizes as Unzoomed and
starts applying zoom at use time throughout the code. Since FlexBasis is
a superset of PreferredSize, that is also marked as unzoomed here.

Much of this patch is going through the various places, which is usually
layout code, and applying the used zoom value of the box we are
evaluating the property of. For example, if we need to evaluate the
width property of a box, we will need to supply its used zoom value. I
will not call out every single time this is done in the commit message
below, since that would mean I would be saying that we are just applying
the box&apos;s zoom value when evaluating its size, but I will call out some
interesting and notable cases where it is not completely as
straightforward as that.

* LayoutTests/TestExpectations:
* LayoutTests/fast/css/viewport-units-zoom.html:
I am disabling this test for now with this feature flag on since it is
incorrect with regards to CSS Zoom. I am currently working on rewriting
this test as part of <a href="https://bugs.webkit.org/show_bug.cgi?id=301404">https://bugs.webkit.org/show_bug.cgi?id=301404</a>

* Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
(WebCore::Layout::FormattingGeometry::computedHeightValue const):
Since this lambda walks up the containing block chain and potentially
returns the logical height of some ancestor it needs to be modified to
also return the zoom value associated with that ancestor too.

* Source/WebCore/rendering/AutoTableLayout.cpp:
It seems like the AutoTableLayout code will either evaluate sizes that
come from dedicated renderers (e.g. table cells) or from LayoutStruct,
which is a helper struct used to hold data for a column to be used
during layout. The former is fairly straightforward and we just grab
the zoom from the renderer just like much of the other rendering code.
For the latter we choose to store the used zoom associated with the
column in the struct so that it is easily accessible whenever we
evaluate the other sizes stored in it.

(WebCore::AutoTableLayout::recalcColumn):
When we evaluate a size from a cell we can use its zoom value since it
has its own dedicated renderer and style. At other times the code looks
at the data populated in the associated LayoutStruct for the column, so
we can use the usedZoom value that we stored in when this struct was
populated.

(WebCore::AutoTableLayout::fullRecalc):
Populate the LayoutStruct with the used zoom value that will be used
later whenever we evaluate the sizes stored here.

(WebCore::AutoTableLayout::applyPreferredLogicalWidthQuirks const):
(WebCore::AutoTableLayout::calcEffectiveLogicalWidth):
(WebCore::AutoTableLayout::layout):
* Source/WebCore/rendering/AutoTableLayout.h:
Add the used zoom value as part of the struct so that we can use it
whenever we evaluate the logicalWidth or effectiveLogicalWidth that also
lives in it. This struct gets populated with the column&apos;s data as part
of AutoTableLayout::fullRecalc.

* Source/WebCore/rendering/FixedTableLayout.cpp:
FixedTableLayout is basically separated into two distinct pieces with
regards to applying zoom. FixedTableLayout seems to use a list of sizes
which are computed from calcWidthArray and then ultimately ued inside
of layout(). Since it is computed by iterating over the
RenderTableCols, which are not trivially accessible during the code
that uses the widths inside of layout(), we choose to zoom the width
values here. Afterwards, we just use the values that were stored
without applying zoom to them.

* Source/WebCore/rendering/RenderTableCellInlines.h:
(WebCore::RenderTableCell::styleOrColLogicalWidth const):
We needed to refactor this function to return the zoom that needed to
be used by the caller. In the case where we return
logicalWidthFromColumns, we return a scale factor of 1.0f since those
values should already be zoomed.

(WebCore::RenderTextControl::computeIntrinsicLogicalWidths const):
* Source/WebCore/rendering/RenderTheme.cpp:
Much of the theme code worked by computing the size it wanted for the
form control, multiplying it by the used zoom, and then settings that
value on the form control&apos;s RenderStyle. Since we now apply the zoom
value at use time we can no longer do this and must remove the zoom from
the logic that determined the form control size. However, to make sure
this logic does not break when the feature flag is disabled we may need
to multiply the zoom factor in when it is disabled. In other words, any
call sites that directly used style.usedZoom() must first determine
whether evaluationTimeZoomed is enabled or not.

(WebCore::updateSliderTrackPartForRenderer):
(WebCore::RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioStyle const):
(WebCore::RenderTheme::paintSliderTicks):
(WebCore::RenderTheme::adjustSwitchStyle const):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::usedZoomForComputedStyle const):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::conversionDataForStyle):
Small helper to use the right conversion data when we are computing the
em size that is used to determine other dimensions.

(WebCore::RenderThemeCocoa::adjustInnerSpinButtonStyleForVectorBasedControls const):
(WebCore::RenderThemeCocoa::adjustButtonStyleForVectorBasedControls const):
We should not add in the zoom when evaluating fixedValue here since that
result may end up getting placed on the RenderStyle which would result
in a double zoom.

(WebCore::Style::computeNonCalcLengthDouble):
Since conversionData.computedLineHeightForFontUnits will eventually call
into RenderStyle::computeLineHeight, we cannot use it if the line-height
value is fixed since computeLineHeight will use the zoom factor. Instead
we need to manually convert it here since our conversionData should have
the right zoom factor to use.

* Source/WebCore/style/values/sizing/StyleMaximumSize.h:
* Source/WebCore/style/values/sizing/StyleMinimumSize.h:
* Source/WebCore/style/values/sizing/StylePreferredSize.h:
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::requires):
(WebCore::SVGLengthContext::valueForSizeType):
(WebCore::SVGLengthContext::valueForLength):
* Source/WebCore/svg/SVGLengthContext.h:

Canonical link: <a href="https://commits.webkit.org/302241@main">https://commits.webkit.org/302241@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dcfebf888360ddbeb90e2696284e0c0c444b84f2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128469 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/738 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39300 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135863 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79914 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2eb51ba6-866d-446b-ab56-bd842c626b09) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130341 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/682 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/611 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97800 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65707 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/451e2afa-f3fb-4a92-9a6c-c20be09b9157) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131417 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/492 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115102 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78411 "Found 1 new API test failure: WPE/TestWebsiteData:/webkit/WebKitWebsiteData/itp (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e59c503c-75ee-4e73-b628-62d9e264794c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/449 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33210 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79145 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108859 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33693 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138312 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/576 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/546 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106339 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/619 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111443 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106150 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27041 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/482 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29984 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52916 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/632 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/63827 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/522 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/585 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/589 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

vitorroriz commented on Source/WebCore/css/CSSToLengthConversionData.cpp:
```diff
@@ -97,7 +97,7 @@ FloatSize CSSToLengthConversionData::defaultViewportFactor() const
         return { };
 
-    return m_renderView->sizeForCSSDefaultViewportUnits() / 100.0;
+    return (m_renderView->sizeForCSSDefaultViewportUnits() / 100.0);
```
unrelated?

vitorroriz commented on Source/WebCore/layout/formattingContexts/flex/FlexFormattingContext.cpp:
```diff
@@ -89,11 +89,12 @@ FlexLayout::LogicalFlexItems FlexFormattingContext::convertFlexItemsToLogicalSpa
             auto crossAxis = LogicalFlexItem::CrossAxisGeometry { };
 
-            auto propertyValueForLength = [&](auto& propertyValue, auto availableSize) -> std::optional<LayoutUnit> {
+            auto propertyValueForLength = [&]<typename SizeType>(const SizeType& propertyValue, auto availableSize) -> std::optional<LayoutUnit> {
```
why is this necessary now?

sammygill replied:
It is not!

vitorroriz commented on Source/WebCore/layout/integration/flex/LayoutIntegrationFlexLayout.cpp:
```diff
@@ -83,18 +83,18 @@ static inline Layout::ConstraintsForFlexContent constraintsForFlexContent(const
         if (auto percentageHeight = computedValue.tryPercentage()) {
             if (callRendererForPercentValue)
-                return flexContainerRenderer.computePercentageLogicalHeight(*percentageHeight, RenderBox::UpdatePercentageHeightDescendants::No);
+                return flexContainerRenderer.computePercentageLogicalHeight(SizeType { *percentageHeight }, RenderBox::UpdatePercentageHeightDescendants::No);
```
is this needed?

sammygill replied:
Yes unfortunately the compiler gets confused

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -7500,13 +7500,6 @@ fast/canvas/image-buffer-resource-limits.html [ Slow ]
 
 # Standardized CSS zoom tests.
 imported/w3c/web-platform-tests/css/css-viewport/width.html [ ImageOnlyFailure ]
```
Why is this still failing?

sammygill replied:
It isn't! I just missed it from my output I guess

alanbaradlay commented on Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
```diff
@@ -87,33 +87,34 @@ template<FormattingGeometry::HeightType heightType> std::optional<LayoutUnit> Fo
                 }
                 ASSERT_NOT_REACHED();
-                return CSS::Keyword::Auto { };
+                return { CSS::Keyword::Auto { }, Style::ZoomFactor { 1.0f, 1.0f } };
```
shouldn't we still try to resolve against the current device pixel ratio?

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
```diff
@@ -87,33 +87,34 @@ template<FormattingGeometry::HeightType heightType> std::optional<LayoutUnit> Fo
             ASSERT_NOT_IMPLEMENTED_YET();
         } else {
-            auto nonAnonymousContainingBlockLogicalHeight = [&]() -> Style::PreferredSize {
+            auto nonAnonymousContainingBlockLogicalHeightWithZoom = [&]() -> std::pair<Style::PreferredSize, Style::ZoomFactor> {
```
not convinced that the name change adds much value here.

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/layout/formattingContexts/grid/PlacedGridItem.h:
```diff
@@ -67,6 +67,8 @@ class PlacedGridItem {
 
     const GridAreaLines& gridAreaLines() const { return m_gridAreaLines; }
 
+    const Style::ZoomFactor& usedZoom() const { return m_usedZoom; }
```
this feels very random.

sammygill replied:
I agree I don't want this here ultimately but I don't have a better idea for now that I like

alanbaradlay commented on Source/WebCore/rendering/FixedTableLayout.cpp:
```diff
@@ -232,7 +235,7 @@ void FixedTableLayout::layout()
     for (unsigned i = 0; i < nEffCols; i++) {
         if (auto fixedWidth = m_width[i].tryFixed()) {
-            calcWidth[i] = Style::evaluate<float>(*fixedWidth, Style::ZoomNeeded { });
+            calcWidth[i] = Style::evaluate<float>(*fixedWidth, Style::ZoomFactor { 1.0f, 1.0f });
```
???

sammygill replied:
Added a comment

alanbaradlay commented on Source/WebCore/rendering/RenderTableCellInlines.h:
```diff
@@ -50,14 +50,18 @@ inline const BorderValue& RenderTableCell::borderAdjoiningTableStart() const
-    return styleWidth;
+        return { styleWidth, style.usedZoomForLength() };
+    if (RenderTableCol* firstColumn = table()->colElement(col())) {
+        // logicalWidthFromColumns will return a zoomed size so we return a zoom
+        // factor of 1.0 to avoid double zooming.
+        return { logicalWidthFromColumns(firstColumn, styleWidth), Style::ZoomFactor { 1.0f, style.deviceScaleFactor() } };
```
Maybe we can express this in a better way as returning a zoomed size while saying the zoom factor is 1 is somewhat confusing.

alanbaradlay commented on Source/WebCore/rendering/RenderTableCell.cpp:
```diff
@@ -196,7 +196,7 @@ Style::PreferredSize RenderTableCell::logicalWidthFromColumns(RenderTableCol* fi
         }
 
-        colWidthSum += fixedColWidth->resolveZoom(Style::ZoomNeeded { });
+        colWidthSum += fixedColWidth->resolveZoom(Style::ZoomFactor { 1.0f, 1.0f });
```
???

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/AutoTableLayout.cpp:
```diff
@@ -171,9 +171,10 @@ void AutoTableLayout::fullRecalc()
             unsigned effCol = m_table->colToEffCol(currentColumn);
             unsigned span = column->span();
             if (!colLogicalWidth.isAuto() && span == 1 && effCol < nEffCols && m_table->spanOfEffCol(effCol) == 1) {
+                m_layoutStruct[effCol].usedZoom = column->style().usedZoomForLength();
```
how does this work with column spanning where a cell spans multiple columns that each have different zoom values?

sammygill replied:
We seem to match other engines in that the spanning cell will use its zoom if it's set otherwise it doesn't seem to be affected by zoom. I am not 100% sure how this exactly works from a spec perspective but it seems to match current behavior

alanbaradlay commented on Source/WebCore/rendering/AutoTableLayout.h:
```diff
@@ -58,6 +58,7 @@ class AutoTableLayout final : public TableLayout {
         float effectiveMaxLogicalWidth { 0 };
         float computedLogicalWidth { 0 };
         bool emptyCellsOnly { true };
+        Style::ZoomFactor usedZoom { 1.0f, 1.0f };
```
while 1 as zoom factor makes sense I am not sure if 1 device scale factor should be the default value here.

sammygill replied:
I changed this to just be the zoom factor so that callers are forced to pass in the device scale factor

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -332,7 +332,7 @@ void RenderBlockFlow::computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidth,
 
     if (auto* cell = dynamicDowncast<RenderTableCell>(*this)) {
-        auto tableCellWidth = cell->styleOrColLogicalWidth();
+        auto [ tableCellWidth, usedZoom ] = cell->styleOrColLogicalWidth();
```
usedZoom does not seem to be used in here. Any reason for that?

sammygill replied:
Only because I think I had to but maybe there's a better way (maybe an overload)

alanbaradlay reviewed (commented):
I think many of the mechanical changes (e.g. (e.g. Style::ZoomNeeded { } -> style().usedZoomForLength()) could be moved to a separate PR to make this one more manageable.

pxlcoder commented on Source/WebCore/style/values/primitives/StyleLengthResolution.cpp:
```diff
@@ -283,6 +273,11 @@ double computeNonCalcLengthDouble(double value, CSS::LengthUnit lengthUnit, cons
+            // We can't use computedLineHeightForFontUnits if the line height is fixed since
+            // that will apply the usedZoomFactor. We probably should refactor it so that
+            // does not happen and we don't have to special case this scenario.
+            value *=   Style::evaluate<LayoutUnit>(*fixedLineHeight, Style::ZoomFactor { conversionData.zoom(), conversionData.style()->deviceScaleFactor() }).toFloat();
```
```suggestion
            value *= Style::evaluate<LayoutUnit>(*fixedLineHeight, Style::ZoomFactor { conversionData.zoom(), conversionData.style()->deviceScaleFactor() }).toFloat();
```

pxlcoder commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -3390,10 +3396,10 @@ static void paintSliderTicksForVectorBasedControls(const RenderElement& box, con
+    const float sliderThumbHeightForLayout = isVertical ? kDefaultSliderThumbWidth : kDefaultSliderThumbHeight;
 
-    const auto usedZoom = style.usedZoom();
+    const auto usedZoom =  usedZoomForComputedStyle(style);
```
```suggestion
    const auto usedZoom = usedZoomForComputedStyle(style);
```

lilyspiniolas commented on Source/WebCore/html/SearchInputType.cpp:
```diff
@@ -200,13 +200,15 @@ float SearchInputType::decorationWidth(float) const
         // FIXME: Document what invariant holds to allow only using fixed logical widths?
-        if (auto fixedLogicalWidth = resultsButton->renderStyle()->logicalWidth().tryFixed())
-            width += fixedLogicalWidth->resolveZoom(Style::ZoomNeeded { });
+        auto* renderStyle = resultsButton->renderStyle();
```
```suggestion
        CheckedPtr renderStyle = resultsButton->renderStyle();
```

pxlcoder replied:
`CheckedPtr`?

lilyspiniolas commented on Source/WebCore/html/SearchInputType.cpp:
```diff
@@ -200,13 +200,15 @@ float SearchInputType::decorationWidth(float) const
         // FIXME: Document what invariant holds to allow only using fixed logical widths?
-        if (auto fixedLogicalWidth = cancelButton->renderStyle()->logicalWidth().tryFixed())
-            width += fixedLogicalWidth->resolveZoom(Style::ZoomNeeded { });
+        auto* renderStyle = cancelButton->renderStyle();
```
```suggestion
        CheckedPtr renderStyle = cancelButton->renderStyle();
```

pxlcoder replied:
`CheckedPtr`?

nt1m commented on Source/WebCore/layout/formattingContexts/FormattingGeometry.h:
```diff
@@ -78,8 +78,9 @@ class FormattingGeometry {
     std::optional<LayoutUnit> computedValue(const auto&, LayoutUnit containingBlockWidth) const;
     std::optional<LayoutUnit> computedValue(const auto&, LayoutUnit containingBlockWidth, const Style::ZoomFactor&) const;
+
+    std::optional<LayoutUnit> fixedValue(const auto&, const Style::ZoomFactor& usedZoom) const;
```
Nit: param name not needed

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html:
```diff
@@ -32,6 +32,10 @@
       "otherValues": ["drop-shadow(rgb(0, 0, 0) 10px 10px 10px)"]
     },
+    "height": {
+      "value": "9x",
```
9px

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html:
```diff
@@ -45,6 +49,14 @@
       "value": "-10px 20px 30px -40px",
       "otherValues": ["-20px 40px 60px -80px"]
     },
+    "min-height" : {
+      "value": "9px",
+      "otherValues": ["18px"]
+    },
+    "min-width" : {
+      "value": "9px",
+      "otherValues": ["18px"]
+    },
```
```suggestion
    "min-height": {
      "value": "9px",
      "otherValues": ["18px"]
    },
    "min-width": {
      "value": "9px",
      "otherValues": ["18px"]
    },
```

Can you also add max as well?

sammygill committed:
[CSS Zoom] Apply zoom to Preferred, Minimum, and Maximum sizes when evaluating
https://bugs.webkit.org/show_bug.cgi?id=300052
rdar://161848512

Reviewed by Tim Nguyen.

This patch marks the Preferred, Minimum, Maximum sizes as Unzoomed and
starts applying zoom at use time throughout the code. Since FlexBasis is
a superset of PreferredSize, that is also marked as unzoomed here.

Much of this patch is going through the various places, which is usually
layout code, and applying the used zoom value of the box we are
evaluating the property of. For example, if we need to evaluate the
width property of a box, we will need to supply its used zoom value. I
will not call out every single time this is done in the commit message
below, since that would mean I would be saying that we are just applying
the box's zoom value when evaluating its size, but I will call out some
interesting and notable cases where it is not completely as
straightforward as that.

* LayoutTests/TestExpectations:
* LayoutTests/fast/css/viewport-units-zoom.html:
I am disabling this test for now with this feature flag on since it is
incorrect with regards to CSS Zoom. I am currently working on rewriting
this test as part of https://bugs.webkit.org/show_bug.cgi?id=301404

* Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
(WebCore::Layout::FormattingGeometry::computedHeightValue const):
Since this lambda walks up the containing block chain and potentially
returns the logical height of some ancestor it needs to be modified to
also return the zoom value associated with that ancestor too.

* Source/WebCore/rendering/AutoTableLayout.cpp:
It seems like the AutoTableLayout code will either evaluate sizes that
come from dedicated renderers (e.g. table cells) or from LayoutStruct,
which is a helper struct used to hold data for a column to be used
during layout. The former is fairly straightforward and we just grab
the zoom from the renderer just like much of the other rendering code.
For the latter we choose to store the used zoom associated with the
column in the struct so that it is easily accessible whenever we
evaluate the other sizes stored in it.

(WebCore::AutoTableLayout::recalcColumn):
When we evaluate a size from a cell we can use its zoom value since it
has its own dedicated renderer and style. At other times the code looks
at the data populated in the associated LayoutStruct for the column, so
we can use the usedZoom value that we stored in when this struct was
populated.

(WebCore::AutoTableLayout::fullRecalc):
Populate the LayoutStruct with the used zoom value that will be used
later whenever we evaluate the sizes stored here.

(WebCore::AutoTableLayout::applyPreferredLogicalWidthQuirks const):
(WebCore::AutoTableLayout::calcEffectiveLogicalWidth):
(WebCore::AutoTableLayout::layout):
* Source/WebCore/rendering/AutoTableLayout.h:
Add the used zoom value as part of the struct so that we can use it
whenever we evaluate the logicalWidth or effectiveLogicalWidth that also
lives in it. This struct gets populated with the column's data as part
of AutoTableLayout::fullRecalc.

* Source/WebCore/rendering/FixedTableLayout.cpp:
FixedTableLayout is basically separated into two distinct pieces with
regards to applying zoom. FixedTableLayout seems to use a list of sizes
which are computed from calcWidthArray and then ultimately ued inside
of layout(). Since it is computed by iterating over the
RenderTableCols, which are not trivially accessible during the code
that uses the widths inside of layout(), we choose to zoom the width
values here. Afterwards, we just use the values that were stored
without applying zoom to them.

* Source/WebCore/rendering/RenderTableCellInlines.h:
(WebCore::RenderTableCell::styleOrColLogicalWidth const):
We needed to refactor this function to return the zoom that needed to
be used by the caller. In the case where we return
logicalWidthFromColumns, we return a scale factor of 1.0f since those
values should already be zoomed.

(WebCore::RenderTextControl::computeIntrinsicLogicalWidths const):
* Source/WebCore/rendering/RenderTheme.cpp:
Much of the theme code worked by computing the size it wanted for the
form control, multiplying it by the used zoom, and then settings that
value on the form control's RenderStyle. Since we now apply the zoom
value at use time we can no longer do this and must remove the zoom from
the logic that determined the form control size. However, to make sure
this logic does not break when the feature flag is disabled we may need
to multiply the zoom factor in when it is disabled. In other words, any
call sites that directly used style.usedZoom() must first determine
whether evaluationTimeZoomed is enabled or not.

(WebCore::updateSliderTrackPartForRenderer):
(WebCore::RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioStyle const):
(WebCore::RenderTheme::paintSliderTicks):
(WebCore::RenderTheme::adjustSwitchStyle const):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::usedZoomForComputedStyle const):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::conversionDataForStyle):
Small helper to use the right conversion data when we are computing the
em size that is used to determine other dimensions.

(WebCore::RenderThemeCocoa::adjustInnerSpinButtonStyleForVectorBasedControls const):
(WebCore::RenderThemeCocoa::adjustButtonStyleForVectorBasedControls const):
We should not add in the zoom when evaluating fixedValue here since that
result may end up getting placed on the RenderStyle which would result
in a double zoom.

(WebCore::Style::computeNonCalcLengthDouble):
Since conversionData.computedLineHeightForFontUnits will eventually call
into RenderStyle::computeLineHeight, we cannot use it if the line-height
value is fixed since computeLineHeight will use the zoom factor. Instead
we need to manually convert it here since our conversionData should have
the right zoom factor to use.

* Source/WebCore/style/values/sizing/StyleMaximumSize.h:
* Source/WebCore/style/values/sizing/StyleMinimumSize.h:
* Source/WebCore/style/values/sizing/StylePreferredSize.h:
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::requires):
(WebCore::SVGLengthContext::valueForSizeType):
(WebCore::SVGLengthContext::valueForLength):
* Source/WebCore/svg/SVGLengthContext.h:

Canonical link: https://commits.webkit.org/302241@main
